rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,238 @@
+
+When the DNS Operator reconciles a DNS Record (kuadrantRecord) it will first look to the referenced providerSecret to see if a multiClusterID is set, if so the operator will first pull the most recent records from the DNS Provider and construct a separate DNS Record which accurately reflects the state of the zone in the provider (zoneRecord).
+
+As this could result in many requests to the DNS Provider API a mechanism is required to reduce this to only request when required:
+
+#### Update GUID Logic
+
+Currently the DNS Operator builds the LB CNAME with a guid based on the name of the gateway, as this could potentially be different on different clusters, this will need to be updated to generate based on the zone's root hostname, to ensure it is consistently generated on all clusters that interact with this zone.
+
+#### Testing if a DNS Provider pull is Required
+
+When the DNS Operator publishes any changes to a zone, it will also add/update a (`kuadrant.last_modified_at`) TXT record with the current timestamp.
+
+If the DNS Operator has no local zoneRecords for a zone (in a providerSecret with a multiClusterID), it will list and build the zoneRecords.
+
+If the DNS Operator already has zoneRecords it will first execute a GET request for the `kuadrant.last_modified_at` TXT record, if the value of that field is more recent than the zoneRecords lastModified time, then the DNS Operator will list and rebuild it's zoneRecords.
+
+#### Ensure local records are accurate
+
+Once the zoneRecord is created / updated on cluster, the operator will then scan through this DNS Record and remove any records that contain the local multiClusterID stored in the status of the DNS Record (more info on this later). Then inject the values from the kuadrantRecord into the zoneRecord. After this the DNS Operator will sanity check the result and prune any bad records (we will expand on that later). Once the zoneRecord is updated and sanity checked, if it has changed then the result is sent back to the DNS Provider so that the zone can be updated and made accurate (with an added update to the (`kuadrant.last_modified_at`) TXT record.
+
+#### Update response to failing health check 
+
+When a health check fails, there are 2 scenarios that need to be considered:
+
+##### Local Health Check Failing
+
+In this scenario, the cluster will immediately remove the IP from the DNS Record unless it is the only remaining IP, as it is a local workload that is not functional.
+
+##### Remote Health Check Failing
+
+After ensuring its own hashed multiClusterID is present in the relevant TXT record, the controller will count the totalClusters which is the sum of unique multiClusterIDs in that hosts A records. Then it will count the reportingClusters which is the number of multiClusterIDs in the relevant TXT record, if the reportingClusters/totalClusters >= HealthCheck.MultiClusterFailureThreshold then this IP will be removed for this host.
```

Comment:
Can two clusters do the removal at the same time thus failing the `MultiClusterFailureThreshold` constraint?

Reply:
although 2 clusters could try to remove a bad A/CNAME record at the same time, one will succeed and the other will fail (as the record no longer exists), the removal of bad IPs does not impact the `failureThreshold`.

Reply:
is it possible for them to choose a different record?

Reply:
do you mean by mistake? Then no.